### PR TITLE
initial pty support

### PR DESCRIPTION
**Pull request by grondo**
This PR introduces the *simple* pty support for Flux jobs mentioned in #282 (wow has it really been 5 years). It started out as an attempt to tunnel a basic terminal session manager like [`dtach`](https://github.com/crigler/dtach) or [`abduco`](https://github.com/martanne/abduco) over flux message endpoints, but it ended up being much easier to implement a clone of these utilities, rather than implementing something like transparent Unix domain socket tunneling.

At some point in the future we can implement a more featured solution using a per-job ssh and/or tmux session to get perfect pty support, but for now this interim solution should work for the 90% use case.

### TL;DR

Try:
```
$ flux mini run -o pty bash
```
also
```
$ flux terminus start
^\ # (detach)
$ flux terminus list
server at 1000-terminus running on rank 0 since Tue Apr  7 01:01:32 2020
1 current session:
0: [/bin/bash] 0 clients (created Tue Apr  7 01:01:32 2020)
$ flux terminus attach 0
```

The PR has 3 main parts, see below. The PR is still a wip while I try to figure out a strategy for adding reasonable tests. Also please give the new facilities a try and give feedback, I've only tested on my local system with one terminal environment. (A weak point here may be restoring terminal settings on the client side. We may have to eventually link with ncurses to get more robust in this area)

Still TODO:

 - [ ] tests (including memleak tests)
 - [ ] look into batching tty updates (tty is in character-at-a-time mode, so this might currently generate a lot of messages)
 - [ ] add a mode or option for `flux job attach` to ignore pty option (w/a caveats when attaching to pty)
 - [ ] consider if we want/need a mode where more than just rank 0 gets a pty. We could also start a "terminus" server in the shell to allow users to "log in" to their job (?)

### libterminus 

Support for remote ptys and terminal session management is introduced in a new library `libterminus` (*termin*al *u*ser *s*ervices (or *s*essions or *s*omething)), which implements the following features:

 * `pty.c`: pseudoterminal multiplexer as a service -- like `dtach` but uses flux service and messages instead of Unix domain socket. Supports multiple clients, attach and detach. See comment at top of file for remote pty protocol (it is simple).
 * `client.c`: client side code for attaching to pty service. This contains the setup code to put the local terminal in raw mode, as well as process the detach character (`^\`), and handle window resize events. No other special processing of terminal data is done.
 * `terminus.[ch]`:  wraps the pty server class in an outer service which allows management of multiple terminal "sessions" under a common service endpoint. Clients can request new sesssions, attach to sessions, kill sessions etc.

### flux-terminus

A new command `flux-terminus` is also introduced which uses `libterminus` to implement a user interface to the `terminus` service. Originally this started as a simple test program, but it seemed like it might be useful or at least a bit entertaining as a flux command. 

`flux-terminus` offers simple terminal session management similar to `dtach`, but since it is using flux services and messages instead of a Unix domain socket, you can start sessions on one broker rank and attach from any other rank within the instance.

To start a new session try, `flux terminus start`. If no `flux-terminus` server is already running this will cause the utility to fork itself into the background and register on service `<userid>-terminus` by default. The client will then attach to the newly created session. You can detach with the detach character (Currently `^\`), then reattach later with `flux terminus attach 0`. Multiple sessions can be created and listed with `flux terminus list`.

### shell pty plugin

For resolution of #282, this PR introduces a very simple `pty` plugin for the shell. If the "pty" shell option is enabled, the plugin creates a pty server for rank 0 *only* at shell service method "pty.0", and drops the pty service name into the `shell.init` event context.

On the `flux-job attach` side, if there is a pty enabled, we create a new pty client and attach it to the remote instead of watching the normal stdio. Please see the caveats in the commit that introduces this support.

As a demo try `flux mini run -o pty bash` or `flux mini run -o pty flux start`.

### CODE REVIEW

**Comment by grondo** [comment]
Another bit here that might need explanation is the new `libutil/llog.h` header. I was developing something similar that would allow log messages to be handled appropriately in the various contexts in which `libterminus` would be used (broker service, utility, and shell plugin), and happened to run across this header-only solution I found to be pretty neat. I incorporated it essentially with minor modifications, though we could easily just do our own, simpler thing if the current header is too much (or the license of the original project is incompatible (it was MIT I think))

**Comment by grondo** [comment]
Sorry, going to force-push to fixup some whitespace errors and a couple nits found by lgtm and travis-ci.

File: src/common/libutil/llog.h [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* Flux internal library logging interface.
+ *
+ * Mostly taken from libtsm shl-llog.h, with updates for Flux
+ * typdef naming styles.
```
Comment (garlick) [inline src/common/libutil/llog.h:14]: just noticed spelling error "typdef"

Comment (grondo) [reply src/common/libutil/llog.h:14]: Thanks!

**Comment by garlick** [comment]
I've poked at this a little bit this morning and can't find anything that doesn't work as I would have expected!  Beautiful job mapping this service onto the flux API.

It seems like there are lots of interesting possibilities for building on this - but just being able to start an interactive shell within a running instance is a huge usability improvement!

Brilliant!

**Comment by garlick** [comment]
> We could also start a "terminus" server in the shell to allow users to "log in" to their job 

That sounds intriguing.  Right now, the `flux job attach` is the only way to connect with the pty of a running job, correct?   If you had terminus servers running with each job (within the job shell, is that what yo meant?), would there be a way to see them all with `flux terminus list`?   (I'm asking just to improve my understanding, not because I think this PR should go there)

**Comment by grondo** [comment]
> Right now, the flux job attach is the only way to connect with the pty of a running job, correct?

You could also attach with:
```console
$ flux terminus attach --service=+<shell_service> --rank=<leader_rank> 0
```

However, note the `+` prefix required to force `flux-terminus` to drop the `-terminus` suffix it usually adds to the provided `service` argument. (I agree this is kind of wonky - it is required because the terminus server library listens at `*terminus.*`, but I should probably try to fix that...)

> If you had terminus servers running with each job (within the job shell, is that what yo meant?), would there be a way to see them all with flux terminus list? (I'm asking just to improve my understanding, not because I think this PR should go there)

Not really - there isn't any real "discovery" of the terminus service. The way `flux-terminus` currently works is that it looks for a server listening at `<userid>-terminus` on the local rank by default (er actually `FLUX_NODEID_ANY`). To list current sessions it just pokes the `.list` service on that server.

If we ran a terminus server in each shell, then `flux-terminus list` would have to be redirected to the service and rank of each shell, e.g. `flux terminus list -r 42 -s 1000-shell-123456789.terminus`, and you couldn't (for now) list all the sessions across a job at once. However, we could easily add a layer to `flux-terminus` that had knowledge of flux jobs, but I wonder if the more useful item of work would be the subprocess server described in #2298.

File: src/common/libterminus/terminus.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+typedef void (*flux_terminus_server_empty_f) (struct flux_terminus_server *ts,
+                                              void *arg);
+
+
+struct flux_terminus_server * flux_terminus_server_create (flux_t *h);
+
+void flux_terminus_server_destroy (struct flux_terminus_server *ts);
+
+
+/*  Set internal libterminus loggig function to 'log_fn'. If unset,
```
Comment (chu11) [inline src/common/libterminus/terminus.h:43]: typo "loggig"

File: src/common/libterminus/terminus.c [hunk 3]
```diff
@@ -0,0 +1,658 @@
...
+ *   OUT:  {}
+ *
+ * - Kill all sessions: *terminus.kill-server
+ *
+ *   IN:   {}
+ *   OUT:  {} (response after all sessions exit)
+ *
+ * Sessions are managed on *terminus.ID service endpoints.
+ * Once the session ID is known, a client may connect directly to
+ * to the pty server at this service, usint the protocol detailed
```
Comment (chu11) [inline src/common/libterminus/terminus.c:42]: typo, usint?

File: src/common/libterminus/pty.c [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+}
+
+int flux_pty_kill (struct flux_pty *pty, int sig)
+{
+    pid_t pgrp = -1;
+    if (ioctl (pty->master, TIOCSIG, sig) >= 0)
+        return 0;
+    if (ioctl (pty->master, TIOCGPGRP, &pgrp) >= 0
+        && pgrp != -1
+        && kill(-pgrp, sig) >= 0)
```
Comment (chu11) [inline src/common/libterminus/pty.c:167]: space betore `(` :-)

File: src/common/libterminus/client.c [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+    c->term.c_cc[VLNEXT] = _POSIX_VDISABLE;
+    c->term.c_cc[VMIN] = 1;
+    c->term.c_cc[VTIME] = 0;
+    if (tcsetattr (STDIN_FILENO, TCSANOW, &c->term) < 0) {
+        llog_warning (c, "failed to setup terminal\n");
+        return -1;
+    }
+    term_needs_restoration = true;
+    atexit (flux_pty_client_restore_terminal);
+    cls ();
```
Comment (chu11) [inline src/common/libterminus/client.c:202]: any particular reason to clear the screen on pty?  I think it's more naturally to just fall into it on the current screen.  But perhaps that's just me.

File: src/cmd/flux-job.c [hunk 6]
```diff
@@ -1726,6 +1727,77 @@ static void finish_mpir_interface ()
...
+    n = snprintf (topic, sizeof (topic), "%s.%s", ctx->service, pty_service);
+    if (n >= sizeof (topic))
+        log_err_exit ("Failed to build pty service topic at %s.%s",
+                      ctx->service, pty_service);
+
+    log_msg ("Attaching to pty at %s", topic);
+    if (flux_pty_client_attach (c, ctx->h, ctx->leader_rank, topic) < 0)
+        log_err_exit ("flux_pty_client_attach");
+
+    if (!(f = flux_pty_client_wait (c))
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/cmd/flux-job.c:1796]: It is very similar to `job_wait()`, in that the future is fulfilled when the pty client completes (e.g. detaches or remote pty process exits). However, this was part of a failed experiment. I was hoping we'd be able to do something like:

 ```c
f = flux_pty_client_wait (client);
if (flux_future_get (f, NULL) < 0)
     log_err_exit ("error waiting for client to complete")
```

i.e. run a pty client in 'now' mode. Howver, this doesn't work currently. Perhaps I need to ensure all the watchers required for the client (fd and signal watchers) are started in the now reactor created by `flux_future_get` in synchronous mode.

It may not be worth it at this time, and perhaps I should just switch to a "exit_notify" mechanism as was done on the pty server side...

File: src/common/libterminus/client.c (hunk 5 above)
Comment (grondo) [reply src/common/libterminus/client.c:202]: Hm, good question. My only answer is that this initially started as a clone of `dtach(1)`, which clears the screen when attaching to a new session. I'm guessing they mostly do this to give a visual cue that you've entered a new session, though I could imagine clearing the screen may help some cases when you put the terminal into raw mode. This also matches tmux/screen behavior so its what users are used to.

However, I just tried without the clear screen and it seems to work fine. Maybe somehow make this optional behavior?

I also learned that xterm has an alternate window buffer we could make use of here (So when you detach/exit tty contents would go back to what was on your screen before attach). I didn't go there now because it seemed superfluous.

**Comment by grondo** [comment]
Pushed a bunch of fixups here while I work on testing, including:

 * Fixes for typos identified above
 * Addition of unit tests for libterminus
 * Fixes and refactoring found during unit test development
 * Fix from @garlick  for libtestutil required for pty server

Still need to write sharness test for `flux-terminus`.

**Comment by grondo** [comment]
I've pushed another checkpoint, comprising a somewhat large set of refactoring and enhancements, unfortunately found to be required while I was developing tests. I'm pushing this now in case anyone wants to poke at it again while I continue working on tests.

Some of these changes required a new set of flags for the client side of the pty, including the ability to do a "synchronous" attach to a remote pty (do not return until the attach response is received), as well as disable raw mode, clear screen, etc., to handle different modes in which a client might be executed.

As a result of refactoring, the shell `pty` plugin now registers a `terminus` server in the shell of every job. When the `-o pty` option is used, rank 0 then gets a pty allocated from this server. e.g.

```console
grondo@asp:~$ flux mini submit -o pty bash
1428129088274432
grondo@asp:~$ flux terminus list -s 1000-shell-1428129088274432.terminus
server at 1000-shell-1428129088274432.terminus running on rank 0 since Wed Apr 15 21:26:01 2020
1 current session:
0: [task0] 0 clients (created Wed Apr 15 21:26:01 2020)
```

This also means you can open an interactive shell on any rank of any job. However, I assume this will largely be just a toy feature and not be all that useful.

@chu11, I also decided you were right and `flux job attach` no longer clear the screen when attaching to a job (but `flux terminus attach` does).

I also found  that there was a race in the 'attach' protocol, since for both jobs and terminus sessions, pty sessions are created asynchronously with the attach request. If a session exits before the attach request is received, the message is dropped and the client will hang. Therefore a new "wait" flag was added to sessions, which causes the server to keep them around until at least one client has attached (or a kill request is recieved). These sessions will be listed as `(exited)` in `flux terminus list`:

```console
server at 1000-terminus running on rank 0 since Wed Apr 15 20:59:01 2020
1 current session:
0: [hostname] (exited) 0 clients (created Wed Apr 15 20:59:01 2020)
```

On first attach the session will immediately exit.

There is still a little work to do in the job-shell pty service to close this race. 
There still seems to be a small window where `flux job attach` can send an `attach` request, which gets dropped instead of ENOSYS, and `flux job attach` will hang indefinitely. One idea would be to have jobs with `-o pty` require that one client attach to the pty before the job will exit (i.e. take a shell completion reference). Perhaps this could just be done for interactive jobs somehow.

**Comment by grondo** [comment]
Hm, some of the new unit tests are failing in Travis. I'll take a look.

**Comment by garlick** [comment]
Having a look at this again.  So far so good!

I like the llog.h logging approach.   If you've ruminated on how we might extend llog.h into other parts of flux, please feel free to open an issue and expound :-).  It seems like implementing a `flux_set_log()` or similar in libflux and then providing logger implementations for:  1) logging to the broker ring buffer (like flux_log),  2) logging to stderr (log log_err), and 3) something like your tap logger would be cool!

**Comment by grondo** [comment]
> If you've ruminated on how we might extend llog.h into other parts of flux, please feel free to open an issue and expound :-). 

Sadly, no. At first, `llog.h` seemed most useful for internal libraries that might be used in one of several contexts, e.g. in flux-shell, in a command line utility, or possibly as a flux service.

**Comment by grondo** [comment]
Ok, this is getting closer.

I've just pushed a bunch of fixups enhancing the testing to try to get coverage up. I also noticed that the task.exec callbacks in the shell were not showing as covered, so I tried adding an explicit call to `__gcov_flush()` in the shell's task pre-exec callback to supplement the one we already have in `libsubprocess/local.c` -- perhaps this will help show the correct coverage there.

Also removed support for `FLUX_TERMINUS_SERVER` since it wasn't really adding much value, and causes problems when starting a new instance.

Finally, added a disconnect handler for libterminus (along with a test), to ensure that clients that disconnect without a `detach` message are removed from ptys.

Let me know when it would be ok to rebase and squash the fixup commits.

**Comment by grondo** [comment]
I pushed one last fix for use of an invalid service topic string in `flux-terminus`.

Any objections if I do a rebase/squash on this branch?

**Comment by grondo** [comment]
Ok, squashed and rebased. I removed `wip` as this is ready for a more formal review.

File: src/common/libterminus/terminus.c [hunk 7]
```diff
@@ -0,0 +1,856 @@
...
+static int start_msghandlers (struct flux_terminus_server *ts,
+                              const struct flux_msg_handler_spec tab[])
+{
+    int i;
+    char topic[128];
+    int len = sizeof (topic);
+    struct flux_match match = FLUX_MATCH_REQUEST;
+    int count = msg_handler_count (tab);
+
+    ts->handlers = calloc (count+1, sizeof (*ts->handlers));
```
Comment (garlick) [inline src/common/libterminus/terminus.c:785]: check for NULL return

File: src/shell/pty.c [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+#include <termios.h>
+
+#include <flux/core.h>
+#include <flux/shell.h>
+
+#include "src/common/libterminus/pty.h"
+#include "src/common/libterminus/terminus.h"
+#include "builtins.h"
+
+static void shlog (void *arg,
```
Comment (garlick) [inline src/shell/pty.c:28]: great sounding function 👍

File: src/common/libutil/llog.h [hunk 9]
```diff
@@ -19,6 +19,36 @@
  *
  * Copyright (c) 2010-2013 David Herrmann <dh.herrmann@gmail.com>
  * Dedicated to the Public Domain
+ *
```
Comment (garlick) [inline src/common/libutil/llog.h:22]: Rebase error?  (llog.h copyright update belongs in a different commit)

File: src/common/libtestutil/util.c [hunk 10]
```diff
@@ -313,6 +313,15 @@ static flux_t *test_connector_create (const char *shmem_name,
     }
     if (!(tcon->h = flux_handle_create (tcon, &handle_ops, flags)))
         BAIL_OUT ("flux_handle_create");
+    /* Allow server to have children
+     */
```
Comment (garlick) [inline src/common/libtestutil/util.c:317]: who are you to say who can and can't have children?  Typical male...

File: src/common/libterminus/client.c [hunk 11]
```diff
@@ -0,0 +1,498 @@
...
+    if (!c || !fn) {
+        errno = EINVAL;
+        return -1;
+    }
+    w = calloc (1, sizeof (*w));
+    if (!w)
+        return -1;
+    w->cb = fn;
+    w->arg = arg;
+    if (zlist_append (c->exit_waiters, w) < 0) {
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libterminus/client.c:123]: Hmm, what errno should we set here? Looking at recent source, `zlist_append()` returns `-1` when `item` is NULL, but asserts on memory allocation failures. Would it be valid to just avoid checking the return code here?

File: src/common/libterminus/client.c [hunk 12]
```diff
@@ -0,0 +1,498 @@
...
+    if (tcgetattr (STDIN_FILENO, &orig_term) < 0)
+        return -1;
+
+    c->term = orig_term;
+
+    cfmakeraw (&c->term);
+    c->term.c_cc[VLNEXT] = _POSIX_VDISABLE;
+    c->term.c_cc[VMIN] = 1;
+    c->term.c_cc[VTIME] = 0;
+    if (tcsetattr (STDIN_FILENO, TCSANOW, &c->term) < 0) {
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libterminus/client.c:219]: Hm, good catch. I'm inclined to include the suggestion (though I've yet to see this done in practice)

File: src/common/libterminus/client.c [hunk 13]
```diff
@@ -0,0 +1,498 @@
...
+}
+
+static void pty_client_exit (struct flux_pty_client *c, flux_future_t *f)
+{
+    const char *message;
+
+    if (flux_rpc_get_unpack (f, "{s:s s:i}",
+                                "message", &message,
+                                "status", &c->wait_status) < 0) {
+        llog_error (c, "rpc unpack: %s", strerror (errno));
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libterminus/client.c:309]: Good point. Should actually be using `future_strerror()` which will also return the future error string if one is set, and fall back to `flux_strerror(3)` which I always forget about.

File: src/common/libterminus/client.c [hunk 14]
```diff
@@ -0,0 +1,498 @@
...
+        return;
+    }
+    if (strcmp (type, "attach") == 0)
+        pty_client_attached (c);
+    else if (strcmp (type, "data") == 0)
+        pty_client_data (c, f);
+    else if (strcmp (type, "resize") == 0)
+        pty_client_resize (c);
+    else if (strcmp (type, "exit") == 0)
+        pty_client_exit (c, f);
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libterminus/client.c:337]: This is a streaming response so the EPROTO wouldn't go anywhere. It should at least be logged, but should the client be torn down as well, or just logged and ignored?

File: src/common/libutil/llog.h (hunk 9 above)
Comment (grondo) [reply src/common/libutil/llog.h:22]: oh, dang... :angry:

File: src/common/libterminus/client.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (garlick) [reply src/common/libterminus/client.c:123]: In other parts of the code base we set ENOMEM, hoping that czmq's oom handling will become configurable at a later date.   (Had a bug open upstream, but their closer-bot killed it)

File: src/common/libterminus/client.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (garlick) [reply src/common/libterminus/client.c:337]: Seems like it should take the same code path as the EPROTO decode error from unpack?

File: src/common/libterminus/pty.c [hunk 15]
```diff
@@ -0,0 +1,593 @@
...
+static struct pty_client *pty_client_create (const flux_msg_t *msg)
+{
+    char *uuid = NULL;
+    struct pty_client *c = NULL;
+
+    if (flux_msg_get_route_first (msg, &uuid) < 0)
+        return NULL;
+    if (!(c = calloc (1, sizeof (*c))))
+        return NULL;
+    c->req = flux_msg_copy (msg, false);
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libterminus/pty.c:103]: Yeah, I think I meant to use `flux_msg_incref()`

File: src/common/libterminus/pty.c [hunk 16]
```diff
@@ -0,0 +1,593 @@
...
+
+    /*  New session so we can attach this process to tty */
+    (void) setsid ();
+
+    /*  Make the slave pty our controlling terminal */
+    if (ioctl (fd, TIOCSCTTY, NULL) < 0)
+        return -1;
+
+    /*  dup pty/in/out onto tty fd */
+    dup2 (fd, 0);
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libterminus/pty.c:317]: Yes, probably. Sorry I was lazy here.

File: src/common/libterminus/pty.c [hunk 17]
```diff
@@ -0,0 +1,593 @@
...
+    return -1;
+}
+
+static int pty_client_set_mode (struct flux_pty *pty,
+                                struct pty_client *c,
+                                const flux_msg_t *msg)
+{
+    const char *mode = NULL;
+    if (flux_msg_unpack (msg, "{s:s}", "mode", &mode) < 0
+        || mode == NULL)
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libterminus/pty.c:426]: Oops, it was initially optional. Sorry about that.

File: src/common/libterminus/terminus.c [hunk 18]
```diff
@@ -0,0 +1,856 @@
...
+        goto error;
+    if (terminus_msg_handler_start (s) < 0)
+        goto error;
+    if (zlist_append (ts->sessions, s) < 0)
+        goto error;
+    if (!zlist_freefn (ts->sessions,
+                       s,
+                       (zlist_free_fn *) terminus_session_destroy,
+                       true))
+        goto error;
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libterminus/terminus.c:220]: Hm, what errno? Should only fail if the item isn't found on the list (just felt wrong not to check return code)

File: src/common/libterminus/terminus.c [hunk 19]
```diff
@@ -0,0 +1,856 @@
...
+        return -1;
+    return 0;
+}
+
+static void terminus_pty_attach (flux_subprocess_t *p, void *arg)
+{
+    struct terminus_session *s = arg;
+    if (flux_pty_attach (s->pty) < 0) {
+        llog_fatal (s->server, "terminus: pty attach: %s\n", strerror (errno));
+        _exit (7);
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libterminus/terminus.c:265]: oh, oops that `7` was some leftover debugging. I'll change it back to `1`.

File: src/common/libterminus/terminus.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libterminus/terminus.c:220]: I set it to `ENOENT` which will be really confusing for a caller, but it should never happen so maybe ok?

File: src/common/libterminus/terminus.c [hunk 20]
```diff
@@ -0,0 +1,856 @@
...
+        FLUX_MSGTYPE_REQUEST,
+        "%s.kill",
+        kill_sessions,
+        FLUX_ROLE_USER
+    },
+    {
+        FLUX_MSGTYPE_REQUEST,
+        "%s.kill-server",
+        kill_server,
+        FLUX_ROLE_USER
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libterminus/terminus.c:756]: Good catch! I could have sworn I did that but I guess only in the actual `pty.c` protocol. Really glad you caught this.

Makes me wonder if we need a special rolemask like `FLUX_ROLE_USER_ONLY` or something that automatically rejects requests that don't match the current userid... not sure if we could make that work in the current msg_handler code or not.

File: src/common/libterminus/client.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libterminus/client.c:225]: I ended up giving up on this. It isn't clear how to check if the settings applied by `cfmakeraw(3)` etc. were successful in a possibly portable way. I also couldn't find any existing example where the terminal settings were checked after `tcsetattr(3)`.

[the remark this answers is not in the document]
Comment (garlick) [reply src/common/libterminus/client.c:225]: Sorry for the wild goose chase.

[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libterminus/client.c:225]: no problem! It was a good observation.

**Comment by grondo** [comment]
Thanks for the thorough review @garlick, I have to apologize for some of the sloppiness!

I believe I've addressed all your comments in the latest fixup commits (plus, found some additional issues). Let me know when you've had a look and it is ok to squash the fixups. Thanks!

**Comment by grondo** [comment]
One builder failed in a test I haven't seen fail before:

```
expecting success: 
        id=$(flux mini submit sleep 600) &&
        flux job wait-event $id start &&
        flux job killall -f &&
	run_timeout 60 flux queue drain

1587339975.324908 start
flux-job: Terminated 1 jobs (0 errors)
not ok 95 - flux job: killall -f kills one job
```

It isn't even clear what exactly failed here. For now I'll just restart the builder.

**Comment by grondo** [comment]
Ok, all green now. Let me know if there's any other changes needed.

**Comment by garlick** [comment]
No - looks good!  Set MWP if ready.  Again, nice work!

**Comment by grondo** [comment]
Let me just test on fluke first and make sure this works on a real system.

**Comment by grondo** [comment]
Ok, verified that this works on fluke:
```console
grondo@fluke108:~/git/flux-core.git$ srun --pty -N16 src/cmd/flux start flux mini run -N2 -n2 -o pty flux start
(flux-9sqbGb) grondo@fluke62:~/git/flux-core.git$ flux hwloc info
2 Machines, 2 Cores, 2 PUs
```

Along with some other sanity checks. Setting MWP.

**Comment by trws** [comment]
Saw this when the close notification for the issue went through, makes me very happy to see the functionality go through, though I admit it threw me to see `libterminus` as it made me assume the `terminus` terminal had a library we were using.

**Comment by grondo** [comment]
>  though I admit it threw me to see libterminus as it made me assume the terminus terminal had a library we were using.

Yeah, sorry I saw that after I had named it.
